src/main/java/software/amazon/awssdk/crt/mqtt/MqttConnection.java

```diff
@@ -127,130 +115,129 @@ public ConnectionState getState() {
+    private static AsyncCallback wrapVoidFuture(CompletableFuture<Void> future) {
+        return new AsyncCallback() {
             @Override
             public void onSuccess() {
-                ack.onSuccess();
+                future.complete(null);
             }
 
             @Override
             public void onFailure(String reason) {
                 Throwable cause = new MqttException(reason);
-                ack.onFailure(cause);
+                future.completeExceptionally(cause);
             }
-        } : null;
-        return callback;
-    }
-
-    public boolean connect(String clientId) {
-        return connect(clientId, true, (short)0, null);
+        };
     }
 
-    public boolean connect(String clientId, MqttActionListener ack) {
-        return connect(clientId, true, (short)0, ack);
-    }
+    private static AsyncCallback wrapPacketFuture(CompletableFuture<Integer> future) {
+        return new AsyncCallback() {
+            @Override
+            public void onSuccess() {
+                future.complete(0);
+            }
 
-    public boolean connect(String clientId, boolean cleanSession) {
-        return connect(clientId, cleanSession, (short)0, null);
+            @Override
+            public void onFailure(String reason) {
+                Throwable cause = new MqttException(reason);
+                future.completeExceptionally(cause);
+            }
+        };
     }
 
-    public boolean connect(String clientId, boolean cleanSession, MqttActionListener ack) {
-        return connect(clientId, cleanSession, (short)0, ack);
+    public CompletableFuture<Void> connect(String clientId) {
+        return connect(clientId, true, 0);
     }
 
-    public boolean connect(String clientId, boolean cleanSession, int keepAliveMs) {
-        return connect(clientId, cleanSession, keepAliveMs, null);
+    public CompletableFuture<Void> connect(String clientId, boolean cleanSession) {
+        return connect(clientId, cleanSession, 0);
     }
 
-    public boolean connect(String clientId, boolean cleanSession, int keepAliveMs, MqttActionListener ack) {
+    public CompletableFuture<Void> connect(String clientId, boolean cleanSession, int keepAliveMs) {
         // Just clamp the keepAlive, no point in throwing
         short keepAlive = (short)Math.max(0, Math.min(keepAliveMs, Short.MAX_VALUE));
-        AsyncCallback connectAck = wrapAck(ack);
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        AsyncCallback connectAck = wrapVoidFuture(future);
         try {
             connectionState = ConnectionState.CONNECTING;
-            mqtt_connect(native_ptr(), clientId, cleanSession, keepAlive, connectAck);
+            mqttConnectionConnect(native_ptr(), clientId, cleanSession, keepAlive, connectAck);
         } catch (CrtRuntimeException ex) {
-            return false;
+            future.completeExceptionally(ex);
         }
```

Comment:
I don't really know the right answer here, but this got me thinking...

This error didn't happen asynchronously, so should we re-throw it as an MqttException? Is that better than stuffing it into the Future that this function just created?

I notice that the `publish()` function will throw if there's an immediate error. So whichever approach we take, let's be consistent.

Reply:
That's a good point. I'll make it consistent: If we get the exception in the immediate call, we'll throw. If not, we'll return it with the future.

Reply:
It's actually more common (and expected) that a method returning CompletableFuture would never throw an exception, and exceptions would always be reported through the future.

Reply:
+1. Having more than one error path is confusing for the customer; see https://github.com/aws/aws-sdk-java-v2/issues/436

Reply:
Ok. Easy fix.